src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,241 @@
+                        .findByUserHandle(result.getUserHandle().getBase64());
+                Set<WebAuthnCredential> credentials = account.getCredentials();
+                Optional<WebAuthnCredential> toUpdate = Optional.empty();
+                ByteArray resultCredentialId = result.getCredentialId();
+                for (WebAuthnCredential c : credentials) {
+                    ByteArray cCredentialId = c.getCredentialId();
+                    if (cCredentialId.equals(resultCredentialId)) {
+                        toUpdate = Optional.of(c);
+                    }
+                }
+                if (toUpdate.isEmpty()) {
+                    return Optional.empty();
+                }
+                WebAuthnCredential credential = toUpdate.get();
+                credentials.remove(credential);
+                credential.setSignatureCount(result.getSignatureCount());
+                credential.setLastUsedOn(new Date());
+                webAuthnCredentialsRepository.save(credential);
+                return Optional.of(account.getUsername());
+            }
+        } catch (Exception e) {
+        }
+        return Optional.empty();
+    }
+
+    UserIdentity getUserIdentityOrGenerate(String username, String realm, Optional<String> displayNameOpt,
+            Optional<Subject> optSub) {
+        String displayName = displayNameOpt.orElse("");
+        Optional<UserIdentity> option = getUserIdentity(username, displayName);
+        if (option.isPresent()) {
+            return option.get();
+        } else {
+            byte[] userHandle = new byte[64];
+            SecureRandom random = new SecureRandom();
+            random.nextBytes(userHandle);
+            final ByteArray userHandleBA = new ByteArray(userHandle);
+            final UserIdentity newUserIdentity = UserIdentity.builder()
+                    .name(username).displayName(displayName)
+                    .id(userHandleBA).build();
+            final WebAuthnUserAccount account = new WebAuthnUserAccount();
+            account.setUsername(username);
+            account.setCredentials(new HashSet<>());
+            account.setRealm(realm);
+            String subject;
+            if (optSub.isPresent()) {
+                subject = optSub.get().getSubjectId();
+            } else {
+                subject = subjectService.generateUuid(SystemKeys.RESOURCE_USER);
+            }
+            account.setSubject(subject);
+            account.setUserHandle(userHandleBA);
+            account.setProvider(provider);
+            webAuthnUserAccountRepository.save(account);
+            return newUserIdentity;
+        }
+    }
+
+    Optional<UserIdentity> getUserIdentity(String username, String displayName) {
+        WebAuthnUserAccount account = webAuthnUserAccountRepository.findByProviderAndUsername(provider, username);
+        if (account == null) {
+            return Optional.empty();
+        }
+        assert (account.getUsername() == username);
```

Comment:
do not use assert in method logic outside of constructors or tests. We don't want to throw assertion failed exceptions at runtime in services. Also, does really exists the case where the username in account differs from the one passed to the findByUsername? If such, the repository is the problem.

Reply:
Removed in ebe3d5f513b0e8034d9f41149805fdb5c9aa1c88

Reply:
Removed in ebe3d5f513b0e8034d9f41149805fdb5c9aa1c88